src/main/java/com/ocient/jdbc/XGConnection.java

```diff
@@ -4349,73 +4368,91 @@ private static void validateConnectionProperties(Properties properties) throws S
+		} else {
+			setParameterCalls.add(() -> setParallelism(0, true));
+		}
+		if (priority != null)
+		{
+			setParameterCalls.add(() -> setPriority(priority, false));
+		} else {
+			setParameterCalls.add(() -> setPriority(0.0, true));
+		}
 
-			if (priorityAdjustFactor != null)
-			{
-				setPriorityAdjustFactor(priorityAdjustFactor, false);
-			} else {
-				setPriorityAdjustFactor(0.0, true);
-			}
+		if (forceExternal){
+			setParameterCalls.add(() -> forceExternal(forceExternal));
+		} else{
+			//there shouldn't ever be a situation where resending params needs to turn this off again
+		}
 
-			if (priorityAdjustTime != null)
-			{
-				setPriorityAdjustTime(priorityAdjustTime, false);
-			} else {
-				setPriorityAdjustTime(0, true);
-			}
+		if (priorityAdjustFactor != null)
+		{
+			setParameterCalls.add(() -> setPriorityAdjustFactor(priorityAdjustFactor, false));
+		} else {
+			setParameterCalls.add(() -> setPriorityAdjustFactor(0.0, true));
+		}
 
-			if (serviceClassName != null)
-			{
-				setServiceClassName(serviceClassName, false);
-			} else {
-				setServiceClassName("", true);
-			}
+		if (priorityAdjustTime != null)
+		{
+			setParameterCalls.add(() -> setPriorityAdjustTime(priorityAdjustTime, false));
+		} else {
+			setParameterCalls.add(() -> setPriorityAdjustTime(0, true));
+		}
 
-		} catch (SQLException e){
-			// This should never happen. We protect against bad arguments for these settings in the driver default. Also, we protect
-			// against them when these methods are called from statement.executeUpdate(...). So the arguments here should always be valid.
-			LOGGER.log(Level.WARNING, String.format("resendParameters go unexpected exception %s with message %s", e.toString(), e.getMessage()));
+		if (serviceClassName != null)
+		{
+			setParameterCalls.add(() -> setServiceClassName(serviceClassName, false));
+		} else {
+			setParameterCalls.add(() -> setServiceClassName("", true));
 		}
+
+		CompletionStage<?> allStages = StageSupport.voidStage();
+
+		for(Callable<CompletionStage<?>> setParameterCall: setParameterCalls) {
+
+			// Loop through all the async function calls that set a parameter.
+			// We compose these async calls because they must be executed serially (not in parallel)
+			// due to our simplex transport model.
+
+			allStages = allStages.thenCompose(r -> {
+				try {
+					return setParameterCall.call();
+				} catch (Exception e) {
+					// This should never happen. We protect against bad arguments for these settings in the driver default. Also, we protect
+					// against them when these methods are called from statement.executeUpdate(...). So the arguments here should always be valid.
+					LOGGER.log(Level.WARNING, String.format("resendParameters got unexpected exception %s with message %s", e.toString(), e.getMessage()));
+					return CompletableFuture.failedFuture(e);
+				}
+			});
+		}
+		
+		return allStages.thenAccept(r -> {});
```

Comment:
(nit)

This is super clever! Love this approach. 

This is a good opportunity to highlight the `AsyncIterator` interface which provides some syntactic sugar for something like this.

```suggestion
		// Supplier instead of Callable so we don't have to handle the exception on the synchronous interface
                return AsyncIterator
                    .toIterator(setParameterSuppliers)
                    .thenCompose(Supplier::get)
                    
                    // Returns a CompletionStage<List<SQLResult<Void>>> so the caller can decide what to do if any tasks failed
                    .collect(Collectors.toList());
```

Reply:
```
// Returns a CompletionStage<List<SQLResult<Void>>> so the caller can decide what to do if any tasks failed
// i.e.
whateverThisMethodIs().thenApply(results -> {
  List<SQLResult<Void>> failedResults = results.stream().filter(SQLResult::isException).collect(Collectors.toList());
  if (failedResults.isEmpty()) {
    return SQLResult.voidResult();
  }
  // TODO add suppressed exceptions
  return failedResults.get(0);
});
```